### ISSUE TITLE
Support cred plugins in the "twistd conch" command line

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4753]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4753|
|Type|enhancement|
|Created|2010-12-03 03:44:09Z|
|Branch|https://github.com/twisted/twisted/tree/conch-cred-plugin-4753|

It would be really great, for all of the standard reasons, to be able to specify the source of credentials material for conch.  This would be extra cool, since it might eventually lead to a conch server that can be run as a user other than root (currently root is required for several reasons, among them that credentials are sometimes checked via PAM or the shadow database).

<details><summary>Searchable metadata</summary>

```
trac-id__4753 4753
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_conch_cred_plugin_4753 branches/conch-cred-plugin-4753
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1291347849000000 1291347849000000
changetime__1331590633000000 1331590633000000
version__None None
owner__cyli cyli

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Should we have a separate ticket for 'allow starting conch server without root privileges'?

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

(In [33042]) Branching to 'conch-cred-plugin-4753'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

(In [33549]) Branching to 'conch-cred-plugin-4753'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

(In [33550]) Add cred plugin to conch
refs [#4753](https://github.com/twisted/twisted/issues/4753)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli commented|
|-|-|

(In [33552]) Fix for python 2.5 (with context not supported)
refs [#4753](https://github.com/twisted/twisted/issues/4753)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

[buildbot results: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-cred-plugin-4753]

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @cyli|
|-|-|

Looks reasonably good, just a couple of things need fixing:

  1. When `--auth` is passed, a user should be able to presume that the authentication mechanisms that they specify are the _only_ mechanisms that will be honored; it might be a security flaw to allow otherwise.  (For example, you might want to run a locked-down server.)  The default set of UNIX, publickey, and maybe PAM should remain the default, for compatibility, but if the user specifies a different mechanism they shouldn't be used.
  2. The default set of authentication mechanisms really needs to be documented, since without that the user doesn't know what they're getting in terms of who can log in.
  3. You can do this in a separate ticket, but the SSH public key and PAM checkers do not have corresponding strcred plugins, which means that the only way to get those features is to rely on the default (i.e. the old bad way, not the new good way).  The relevant plugin machinery is `twisted.cred.strcred` and `ICheckerFactory` plugins, which you can see in `twisted/plugins/cred_*`; it should be pretty easy to add, but smaller tickets are always better.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

One thing to be aware of, perhaps, is that the PAM stuff probably doesn't work, or doesn't work well enough to use outside of a toy setting.  It makes Twisted calls from the wrong thread and exposes an API that almost cannot be implemented correctly.  Trace how `pamConversion` is defined and used and you'll see the problems.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_9):
> One thing to be aware of, perhaps, is that the PAM stuff probably doesn't work, or doesn't work well enough to use outside of a toy setting.  It makes Twisted calls from the wrong thread and exposes an API that almost cannot be implemented correctly.  Trace how `pamConversion` is defined and used and you'll see the problems.

Is there an existing ticket about this?

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli commented|
|-|-|

Does that mean I should not make a PAM strcred plugin?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Replying to [cyli](#note_11):
> Does that mean I should not make a PAM strcred plugin?

Yeah, let's leave that out.  If, as exarkun says, it's making Twisted calls from the wrong thread, it should never be enabled.  You should also make a separate ticket for fixing the existing PAM integration, and mention that that ticket should add a strcred plugin for it.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_13"></a>|@cyli commented|
|-|-|

Note: there seems to be a UNIX Password checker in twisted.conch.checkers.py, as well as a UNIX checker in twisted.cred.checkers.py.  I've opened [#5518](https://github.com/twisted/twisted/issues/5518) about eventually removing this.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli commented|
|-|-|

Sorry, correction - twisted.plugins.cred_unix.py

Replying to [cyli](#note_13):
> Note: there seems to be a UNIX Password checker in twisted.conch.checkers.py, as well as a UNIX checker in twisted.cred.checkers.py.  I've opened [#5518](https://github.com/twisted/twisted/issues/5518) about eventually removing this.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_15"></a>|@cyli commented|
|-|-|

There are already two PAM tickets:

[#3728](https://github.com/twisted/twisted/issues/3728) seems to refer to the PAM issues that exarkun mentioned, and [#2970](https://github.com/twisted/twisted/issues/2970) is a ticket for making a PAM checker factory.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_16"></a>|@cyli commented|
|-|-|

(In [33666]) Default checkers should only be auth methods if --auth paramter isn't set. Also
add SSH cred plugin.  refs [#4753](https://github.com/twisted/twisted/issues/4753)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_17"></a>|@cyli commented|
|-|-|

(In [33667]) Attempts to fix 2 bugs: key error bug for those machines with PAM, and import error when the Crypto module is not available.  refs [#4753](https://github.com/twisted/twisted/issues/4753)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_18"></a>|@cyli commented|
|-|-|

Forgot to include a ticket number in [33668](http://twistedmatrix.com/trac/changeset/33668)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_19"></a>|@cyli commented|
|-|-|

(In [33669]) Attempt to fix key bug in test_tap, and also an attempt to fix doc bug.  refs [#4753](https://github.com/twisted/twisted/issues/4753)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_20"></a>|@cyli commented|
|-|-|

(In [33670]) SSH cred plugin also needs pyasn1 - check for that before running cred plugin test.  refs [#4753](https://github.com/twisted/twisted/issues/4753)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_21"></a>|@cyli commented|
|-|-|

(In [33671]) Remove debug prints.  refs [#4753](https://github.com/twisted/twisted/issues/4753)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_22"></a>|@cyli removed owner|
|-|-|

[buildbot results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-cred-plugin-4753) I don't think the doc builder errors are mine.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_24"></a>|@thijstriemstra commented|
|-|-|

There's a typo in test_strcrd.py: ```skip = "PyCryto is not available"```

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_25"></a>|@cyli commented|
|-|-|

Fixed typo in [33672](http://twistedmatrix.com/trac/changeset/33672)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_26"></a>|@cyli commented|
|-|-|

Meh, lack of tests to make sure that --auth actually overrides the auth methods that seem to be default in SSHUserAuthServer

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_27"></a>|@cyli commented|
|-|-|

After discussion with exarkun, we're going to add support for anonymous, etc. in a different ticket

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun set owner to @cyli|
|-|-|

Thanks for working on this.  I'm really happy to see this feature in conch now. :)

  1. in twisted/conch/tap.py
    1. `Options.longdesc` is defined using funny indentation; can you line up the "s? 
    1. `makeService` should have a docstring
  1. in twisted/conch/test/test_tap.py
    1. `MakeServiceTest.setUp` tries to call file.close, but it's missing some ()s.  Also, it might be better to just use `twisted.python.filepath.FilePath.setContent` to write out this file.
    1. The formatting of the multiline return in `test_authFailure` is a little gross.  Can you split that up into multiple statements?  Also, maybe `TestCase.assertFailure` would be helpful here.
    1. return statement in `test_authSuccess` is also maybe worth splitting up.
  1. in twisted/test/test_strcred.py
    1. The lines that got reformatted to be less than 80 columns don't match how most of Twisted is formatted.  Can you make the 2nd line aligned with the open paren of the function call, or put the newline after the open paren and have all the args on the 2nd line.
  1. The news fragment should follow the style guidelines on http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

There were some other indentation-related changes that I didn't comment on specifically, but it would be nice (but not mandatory for this ticket) to fix them.

These points are all pretty simple, please fix and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_29"></a>|@cyli set status to `closed`|
|-|-|

(In [33696]) Merge conch-cred-plugin-4753: support cred plugins in conch

Author: cyli
Reviewer: glyph, exarkun
Fixes: [#4753](https://github.com/twisted/twisted/issues/4753)

Changed twisted.conch.tap so that twistd conch now supports cred plugins
